DOC: Monkey testing with {shinytest2} (#140)
rstudio/shinytest2: vignettes/monkey-testing.Rmd

  let s = document.createElement('script');
  s.src = 'https://unpkg.com/gremlins.js';
  document.body.appendChild(s);
")
```

This creates a `<script>` tag pointing to the right CDN (content delivery network: optimized server to store libraries) and inserts
it at the end of the body.

To test whether everything worked well, we can dump the DOM, look for the scripts. We should be able to find
gremlins.js and try to call `typeof window.gremlins`, which should return an object:

```r
headless_app$get_html("script", outer_html = TRUE)
headless_app$get_js("typeof window.gremlins")
```

If `undefined` is returned instead, it means something went wrong. This is generally because the JS code
is blocked by the network. In this case, let's see what we can do in the next part.

```{r echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("figures/gremlins-inject.png")
```

### Local way

If you work in a corporate environment chances are you'll end up in this situation. 
A solution is then to store and serve a local copy of the gremlin.js script with `shiny::addResourcePath()`.
Assuming `gremlins.js` are in `inst/js/gremlins.min.js`. Add this to the `app.R` file:

```r

> addResourcePath("gremlins", "inst/js/gremlins.min.js")

Reviewer: ```suggestion
shiny::addResourcePath("gremlins", "inst/js/gremlins.min.js")
```

```

We can subsequently inject the gremlins in the DOM and checks whether everything worked as expected:

```r
headless_app$run_js("
  let s = document.createElement('script');
  s.src = './gremlins/gremlins.min.js';
  document.body.appendChild(s);
")

> plop$get_html("script", outer_html = TRUE)
> plop$get_js("typeof window.gremlins")

Reviewer: ```suggestion
headless_app$get_html("script", outer_html = TRUE)
headless_app$get_js("typeof window.gremlins")
```

```

## Unleash the horde

### A bit about gremlins.js

The workflow is rather simple:

  - We create the horde with `gremlins.createHorde()`.
  - We run the monkey test with `horde.unleash();`. 

```js
const horde = gremlins.createHorde();
horde.unleash();
```

`createHorde()` accepts many species of gremlins capable of handling various events such as
clicks, touch, form filling, scrolling, typing, ... described in the gremlins.js [documentation](https://github.com/marmelab/gremlins.js#advanced-usage).

As a side note, we don't recommend using the `scroller` which sometimes [crashes](https://github.com/rstudio/shinytest2/issues/117) the Chrome instance. 

> If your plots rely on random elements like `rnorm`, it is best practice to setup a seed. 

Reviewer: ```suggestion
If your plots rely on random elements like `rnorm`, it is best practice to setup a `seed`. 
```

> By default, all species will attack in random order with a delay of 10 ms between each event. You can also control the attack strategy to fine tune the global behavior. Finally, if you want more control over what gremlins specie should be doing, you can define a custom specie. 

Reviewer: ```suggestion
By default, all species will attack in random order with a delay of 10 ms between each event. You can also control the attack strategy to fine tune the global behavior. Finally, if you want more control over what your gremlins species should be doing, you can define a custom species.
```

### Practice

#### Blind run

In the following, we run the most basic Monkey test configuration:

```r
headless_app$run_js("gremlins.createHorde().unleash();")
```

Result is shown in the gif below. 

```{r echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("figures/gremlins-attack.gif")
```

#### Optimized testing

Does it really makes sense to invoke all species given we only have a slider? Absolutely not!
In the following code, we only want to test the slider input.

```{r echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("figures/gremlins-slider-handle.png")
```

> The best specie to perform this task is the `toucher`, which is able to randomly move the slider input. The documentation specifies many events such as tap, doubletap, gesture and multitouch. It seems more relevant to apply only gesture, consisting of dragging the slider on the x-axis (note: gesture also consider the y-axis which does not make sense for the slider, which is an gremlins.js limitation. This however won't prevent our test from running). Also, we target the slider handle by its class `irs-handle single` to ensure we don't touch any other element and increase the number of maximum touch from 2 to 200. Not how precious is the HTML inspector, which allows to seamlessly inspect and extract any class or id. The __log__ parameter enables log in the JavaScript console:

Reviewer: ```suggestion
The best species to perform this task is the `toucher`, which is able to randomly move the slider input. The documentation specifies many events such as tap, doubletap, gesture and multitouch. It seems more relevant to apply only gesture, consisting of dragging the slider on the x-axis (note: gesture also consider the y-axis which does not make sense for the slider, which is a gremlins.js limitation. This however won't prevent our test from running). Also, we target the slider handle by its class `irs-handle single` to ensure we don't touch any other element and increase the number of maximum touch from 2 to 200. Note how precious is the HTML inspector, which allows to seamlessly inspect and extract any class or id. The __log__ parameter enables logging in the JavaScript console:
```

```js
horde
  .unleash()
  .then(() => {
    console.log('Gremlins test success')
});
```

If you are running on CI/CD and can't see the Chrome Devtools, you can still display the logs. The `console.log` output will be captured by the `{shinytest2}` logs with `headless_app$get_logs()` allowing you to debug and refine the 
monkey testing script. 

```{r echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("figures/gremlins-logs.png")
```

As Monkey testing lasts about 10s, you often want to take a screenshot of the ongoing attack:

```r
headless_app$run_js("
  const customToucher = gremlins.species.toucher({
    // which touch event types will be triggered
    touchTypes: ['gesture'],
    // Touch only if element has class irs-handle single
    canTouch: (element) => element.className === 'irs-handle single',
    log: true,
    maxTouches: 200
  });
    
  gremlins.createHorde({
    //randomizer: new gremlins.Chance(1234), // repeatable
    species: [customToucher],
    mogwais: [gremlins.mogwais.gizmo()]
  }).unleash().then(() => {
    console.log('Gremlins test success')
  });
")

Sys.sleep(3)

> headless_app$screenshot("gremlins.png")

Reviewer: ```suggestion
headless_app$get_screenshot("gremlins.png")
```

available by default in `{shinytest2}`, you can pass them before instantiating the driver:

```r
chromote::set_chrome_args(
  c(
    chromote::default_chrome_args(),
    # Custom flags: see https://peter.sh/experiments/chromium-command-line-switches/
)
```

Some flags are considered by default, particularly `--no-sandbox` which is apply only on CI/CD (Chrome won't start
without) such that you don't need to worry too much about this.

If you run this script locally, you may add `view = TRUE` to open the Chrome Devtools, which will significantly ease the testing calibration. I highly recommend to create the test protocol locally and then move to CI/CD later when all bugs are fixed. In the below figure, the application is shown on the left side panel. The top-right side panel shows the DOM elements (default) and the bottom-right side panel displays the JavaScript console output. 

```{r echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("figures/gremlins-start.png")
```

## Injecting gremlins.js

The next steps consists of injecting the gremlins.js dependency in the DOM so that we can unleash the horde.

### Easy way
The easiest way is to call:

```r

> headless_app$execute_js("
>   (function() {
>     var s = document.createElement('script');
>     s.src = 'https://unpkg.com/gremlins.js';
>     document.body.appendChild(s);
>   })();"
> )

Reviewer: ```suggestion
headless_app$run_js("
  let s = document.createElement('script');
  s.src = 'https://unpkg.com/gremlins.js';
  document.body.appendChild(s);
")
```

```

This creates a `<script>` tag pointing to the right CDN (content delivery network: optimized server to store libraries) and inserts
it at the end of the body.

To test whether everything worked well, we can dump the DOM, look for the scripts. We should be able to find
gremlins.js and try to call `typeof window.gremlins`, which should return an object:

```r
headless_app$get_html("script", outer_html = TRUE)

> headless_app$execute_js("return typeof window.gremlins")

Reviewer: ```suggestion
headless_app$get_js("typeof window.gremlins")
```

```

If `undefined` is returned instead, it means something went wrong. This is generally because the JS code
is blocked by the network. In this case, let's see what we can do in the next part.

```{r echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("figures/gremlins-inject.png")
```

### Local way

If you work in a corporate environment chances are you'll end up in this situation. 
A solution is then to store and serve a local copy of the gremlin.js script with `shiny::addResourcePath()`.
Assuming `gremlins.js` are in `inst/js/gremlins.min.js`. Add this to the `app.R` file:

```r
addResourcePath("gremlins", "inst/js/gremlins.min.js")
```

We can subsequently inject the gremlins in the DOM and checks whether everything worked as expected:

```r

> headless_app$execute_js("
>   (function() {
>     var s = document.createElement('script');
>     s.src = './gremlins/gremlins.min.js';
>     document.body.appendChild(s);
>   })();"
> )
> plop$get_html("script", outer_html = TRUE)
> plop$execute_js("return typeof window.gremlins")

Reviewer: ```suggestion
headless_app$run_js("
  let s = document.createElement('script');
  s.src = './gremlins/gremlins.min.js';
  document.body.appendChild(s);
")
plop$get_html("script", outer_html = TRUE)
plop$get_js("typeof window.gremlins")
```

```

## Unleash the horde

### A bit about gremlins.js

The workflow is rather simple:

  - We create the horde with `gremlins.createHorde()`.
  - We run the monkey test with `horde.unleash();`. 

```js
const horde = gremlins.createHorde();
horde.unleash();
```

`createHorde()` accepts many species of gremlins capable of handling various events such as
clicks, touch, form filling, scrolling, typing, ... described in the gremlins.js [documentation](https://github.com/marmelab/gremlins.js#advanced-usage).

As a side note, we don't recommend using the `scroller` which sometimes [crashes](https://github.com/rstudio/shinytest2/issues/117) the Chrome instance. 

If your plots rely on random elements like `rnorm`, it is best practice to setup a seed. 
By default, all species will attack in random order with a delay of 10 ms between each event. You can also control the attack strategy to fine tune the global behavior. Finally, if you want more control over what gremlins specie should be doing, you can define a custom specie. 

### Practice

#### Blind run

In the following, we run the most basic Monkey test configuration:

```r

> headless_app$execute_js("gremlins.createHorde().unleash();")

Reviewer: ```suggestion
headless_app$run_js("gremlins.createHorde().unleash();")
```

  // which touch event types will be triggered
  touchTypes: ['gesture'],
  // Touch only if element has class irs-handle single
  canTouch: (element) => element.className === 'irs-handle single',
  log: true,
  maxTouches: 200
});
```

We can create our custom horde and disable the FPS and Alert mogwais gremlins since they are not relevant to our case study:

```js
const horde = gremlins.createHorde({
  species: [customToucher],
  mogwais: [gremlins.mogwais.gizmo()]
});
```

As `unleash` is a __promise__, we can execute a function right after, to check whether the script ran as expected.

```js
horde
  .unleash()
  .then(() => {
    console.log('Gremlins test success')
});
```

If you are running on CI/CD and can't see the Chrome Devtools, you can still display the logs. The `console.log` output will be captured by the `{shinytest2}` logs with `headless_app$get_log()` allowing you to debug and refine the 
monkey testing script. 

```{r echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("figures/gremlins-logs.png")
```

As Monkey testing lasts about 10s, you often want to take a screenshot of the ongoing attack:

```r

> headless_app$execute_js(
>   "const customToucher = gremlins.species.toucher({
>     // which touch event types will be triggered
>     touchTypes: ['gesture'],
>     // Touch only if element has class irs-handle single
>     canTouch: (element) => element.className === 'irs-handle single',
>     log: true,
>     maxTouches: 200
>     });
>     
>     gremlins.createHorde({
>       //randomizer: new gremlins.Chance(1234), // repeatable
>       species: [customToucher],
>       mogwais: [gremlins.mogwais.gizmo()]
>     }).unleash().then(() => {
>       console.log('Gremlins test success')
>     });
>   "
> )

Reviewer: ```suggestion
headless_app$run_js("
  const customToucher = gremlins.species.toucher({
    // which touch event types will be triggered
    touchTypes: ['gesture'],
    // Touch only if element has class irs-handle single
    canTouch: (element) => element.className === 'irs-handle single',
    log: true,
    maxTouches: 200
  });
    
  gremlins.createHorde({
    //randomizer: new gremlins.Chance(1234), // repeatable
    species: [customToucher],
    mogwais: [gremlins.mogwais.gizmo()]
  }).unleash().then(() => {
    console.log('Gremlins test success')
  });
")
```
